Debezium heartbeat - crash with "The field specified for partitioning cannot be found in the schema."

This is follow up to https://github.com/memiiso/debezium-server-bigquery/issues/288 (I can't comment on that issue any more) - many thanks for looking into my issue that quickly.  

I tried the latest release but it keeps on crashing with `debezium.source.heartbeat.interval.ms=5000`. The error message is different this time, though:
```
2025-05-02 06:48:07,918 INFO  [io.deb.con.pos.con.PostgresReplicationConnection] (debezium-postgresconnector-tutorial-change-event-source-coordinator) Initializing PgOutput logical decoder publication
2025-05-02 06:48:07,955 INFO  [io.deb.con.pos.con.PostgresConnection] (debezium-postgresconnector-tutorial-change-event-source-coordinator) Obtained valid replication slot ReplicationSlot [active=false, latestFlushedLsn=LSN{B7E/16004100}, catalogXmin=1112898]
2025-05-02 06:48:07,956 INFO  [io.deb.jdb.JdbcConnection] (pool-16-thread-1) Connection gracefully closed
2025-05-02 06:48:07,978 INFO  [io.deb.uti.Threads] (debezium-postgresconnector-tutorial-change-event-source-coordinator) Requested thread factory for component PostgresConnector, id = tutorial named = keep-alive
2025-05-02 06:48:07,979 INFO  [io.deb.uti.Threads] (debezium-postgresconnector-tutorial-change-event-source-coordinator) Creating thread debezium-postgresconnector-tutorial-keep-alive
2025-05-02 06:48:07,979 INFO  [io.deb.con.pos.PostgresStreamingChangeEventSource] (debezium-postgresconnector-tutorial-change-event-source-coordinator) Processing messages
2025-05-02 06:48:08,517 ERROR [io.deb.emb.asy.AsyncEmbeddedEngine] (pool-7-thread-1) Engine has failed with : java.util.concurrent.ExecutionException: com.google.cloud.bigquery.BigQueryException: The field specified for partitioning cannot be found in the schema.
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
        at io.debezium.embedded.async.AsyncEmbeddedEngine.runTasksPolling(AsyncEmbeddedEngine.java:496)
        at io.debezium.embedded.async.AsyncEmbeddedEngine.run(AsyncEmbeddedEngine.java:220)
        at io.debezium.server.DebeziumServer.lambda$start$1(DebeziumServer.java:182)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: com.google.cloud.bigquery.BigQueryException: The field specified for partitioning cannot be found in the schema.
        at com.google.cloud.bigquery.BigQueryRetryHelper.runWithRetries(BigQueryRetryHelper.java:59)
        at com.google.cloud.bigquery.BigQueryImpl.create(BigQueryImpl.java:305)
        at io.debezium.server.bigquery.StreamBigqueryChangeConsumer.createTable(StreamBigqueryChangeConsumer.java:252)
        at io.debezium.server.bigquery.StreamBigqueryChangeConsumer.getTable(StreamBigqueryChangeConsumer.java:263)
        at io.debezium.server.bigquery.StreamBigqueryChangeConsumer.uploadDestination(StreamBigqueryChangeConsumer.java:153)
        at io.debezium.server.bigquery.BaseChangeConsumer.handleBatch(BaseChangeConsumer.java:122)
        at io.debezium.embedded.async.ParallelSmtAndConvertBatchProcessor.processRecords(ParallelSmtAndConvertBatchProcessor.java:56)
        at io.debezium.embedded.async.AsyncEmbeddedEngine$PollRecords.doCall(AsyncEmbeddedEngine.java:1203)
        at io.debezium.embedded.async.AsyncEmbeddedEngine$PollRecords.doCall(AsyncEmbeddedEngine.java:1183)
        at io.debezium.embedded.async.RetryingCallable.call(RetryingCallable.java:47)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        ... 3 more
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
POST https://bigquery.googleapis.com/bigquery/v2/projects/qa-europe-west1-1-2c5eadfd/datasets/qa_ew11/tables?prettyPrint=false
{
  "code": 400,
  "errors": [
    {
      "domain": "global",
      "message": "The field specified for partitioning cannot be found in the schema.",
      "reason": "invalid"
    }
  ],
  "message": "The field specified for partitioning cannot be found in the schema.",
  "status": "INVALID_ARGUMENT"
}
        at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:145)
        at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:118)
        at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:37)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest$3.interceptResponse(AbstractGoogleClientRequest.java:479)
        at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1111)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:565)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:506)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:616)
        at com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc.createSkipExceptionTranslation(HttpBigQueryRpc.java:239)
        at com.google.cloud.bigquery.BigQueryImpl$2.call(BigQueryImpl.java:309)
        at com.google.cloud.bigquery.BigQueryImpl$2.call(BigQueryImpl.java:306)
        at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:102)
        at com.google.cloud.bigquery.BigQueryRetryHelper.run(BigQueryRetryHelper.java:92)
        at com.google.cloud.bigquery.BigQueryRetryHelper.runWithRetries(BigQueryRetryHelper.java:50)
        ... 15 more

2025-05-02 06:48:08,519 INFO  [io.deb.emb.asy.AsyncEmbeddedEngine] (pool-7-thread-1) Engine state has changed from 'POLLING_TASKS' to 'STOPPING'
2025-05-02 06:48:08,521 INFO  [io.deb.con.com.BaseSourceTask] (pool-11-thread-1) Stopping down connector
2025-05-02 06:48:08,994 INFO  [io.deb.jdb.JdbcConnection] (pool-17-thread-1) Connection gracefully closed
2025-05-02 06:48:08,995 INFO  [io.deb.jdb.JdbcConnection] (pool-18-thread-1) Connection gracefully closed
2025-05-02 06:48:08,996 INFO  [io.deb.pip.ChangeEventSourceCoordinator] (debezium-postgresconnector-tutorial-change-event-source-coordinator) Finished streaming
2025-05-02 06:48:08,996 INFO  [io.deb.pip.ChangeEventSourceCoordinator] (debezium-postgresconnector-tutorial-change-event-source-coordinator) Connected metrics set to 'false'
2025-05-02 06:48:08,997 INFO  [io.deb.pip.sig.SignalProcessor] (pool-11-thread-1) SignalProcessor stopped
2025-05-02 06:48:08,998 INFO  [io.deb.ser.DefaultServiceRegistry] (pool-11-thread-1) Debezium ServiceRegistry stopped.
2025-05-02 06:48:08,999 INFO  [io.deb.jdb.JdbcConnection] (pool-19-thread-1) Connection gracefully closed
2025-05-02 06:48:08,999 INFO  [io.deb.emb.asy.AsyncEmbeddedEngine] (pool-7-thread-1) Stopped task #1 out of 1 tasks (it took 478 ms to stop the task).
2025-05-02 06:48:09,002 INFO  [io.deb.emb.asy.AsyncEmbeddedEngine] (pool-7-thread-1) Engine is stopped.
2025-05-02 06:48:09,002 INFO  [io.deb.emb.asy.AsyncEmbeddedEngine] (pool-7-thread-1) Engine state has changed from 'STOPPING' to 'STOPPED'
2025-05-02 06:48:09,002 ERROR [io.deb.ser.ConnectorLifecycle] (pool-7-thread-1) Connector completed: success = 'false', message = 'com.google.cloud.bigquery.BigQueryException: The field specified for partitioning cannot be found in the schema.', error = 'com.google.cloud.bigquery.BigQueryException: The field specified for partitioning cannot be found in the schema.': com.google.cloud.bigquery.BigQueryException: The field specified for partitioning cannot be found in the schema.
        at com.google.cloud.bigquery.BigQueryRetryHelper.runWithRetries(BigQueryRetryHelper.java:59)
        at com.google.cloud.bigquery.BigQueryImpl.create(BigQueryImpl.java:305)
        at io.debezium.server.bigquery.StreamBigqueryChangeConsumer.createTable(StreamBigqueryChangeConsumer.java:252)
        at io.debezium.server.bigquery.StreamBigqueryChangeConsumer.getTable(StreamBigqueryChangeConsumer.java:263)
        at io.debezium.server.bigquery.StreamBigqueryChangeConsumer.uploadDestination(StreamBigqueryChangeConsumer.java:153)
        at io.debezium.server.bigquery.BaseChangeConsumer.handleBatch(BaseChangeConsumer.java:122)
        at io.debezium.embedded.async.ParallelSmtAndConvertBatchProcessor.processRecords(ParallelSmtAndConvertBatchProcessor.java:56)
        at io.debezium.embedded.async.AsyncEmbeddedEngine$PollRecords.doCall(AsyncEmbeddedEngine.java:1203)
        at io.debezium.embedded.async.AsyncEmbeddedEngine$PollRecords.doCall(AsyncEmbeddedEngine.java:1183)
        at io.debezium.embedded.async.RetryingCallable.call(RetryingCallable.java:47)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
POST https://bigquery.googleapis.com/bigquery/v2/projects/qa-europe-west1-1-2c5eadfd/datasets/qa_ew11/tables?prettyPrint=false
{
  "code": 400,
  "errors": [
    {
      "domain": "global",
      "message": "The field specified for partitioning cannot be found in the schema.",
      "reason": "invalid"
    }
  ],
  "message": "The field specified for partitioning cannot be found in the schema.",
  "status": "INVALID_ARGUMENT"
}
        at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:145)
        at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:118)
        at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:37)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest$3.interceptResponse(AbstractGoogleClientRequest.java:479)
        at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1111)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:565)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:506)
        at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:616)
        at com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc.createSkipExceptionTranslation(HttpBigQueryRpc.java:239)
        at com.google.cloud.bigquery.BigQueryImpl$2.call(BigQueryImpl.java:309)
        at com.google.cloud.bigquery.BigQueryImpl$2.call(BigQueryImpl.java:306)
        at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:102)
        at com.google.cloud.bigquery.BigQueryRetryHelper.run(BigQueryRetryHelper.java:92)
        at com.google.cloud.bigquery.BigQueryRetryHelper.runWithRetries(BigQueryRetryHelper.java:50)
        ... 15 more

2025-05-02 06:48:09,014 INFO  [io.deb.ser.DebeziumServer] (main) Received request to stop the engine
2025-05-02 06:48:09,014 INFO  [io.deb.ser.DebeziumServer] (main) Cannot shut down engine now:
2025-05-02 06:48:09,041 INFO  [io.quarkus] (main) debezium-server-bigquery-dist stopped in 0.038s
```

--- a/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/BaseChangeConsumer.java
+++ b/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/BaseChangeConsumer.java
@@ -10,6 +10,7 @@ package io.debezium.server.bigquery;
 
 import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.cloud.bigquery.Schema;
 import io.debezium.DebeziumException;
 import io.debezium.engine.ChangeEvent;
 import io.debezium.engine.DebeziumEngine;
@@ -144,6 +145,18 @@ public abstract class BaseChangeConsumer extends io.debezium.server.BaseChangeCo
       consumerStart = clock.currentTimeInMillis();
       logTimer = Threads.timer(clock, LOG_INTERVAL);
     }
+  }
+
+  protected static boolean schemaHasField(Schema schema, String field) {
+    try {
+      if (schema.getFields() != null && schema.getFields().get(field) != null) {
+        return true;
+      }
+    } catch (Exception e) {
+      return false;
+    }
+
+    return false;
   }
 
   public abstract long uploadDestination(String destination, List<RecordConverter> data);

--- a/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/BaseChangeConsumer.java
+++ b/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/BaseChangeConsumer.java
@@ -147,7 +147,7 @@ public abstract class BaseChangeConsumer extends io.debezium.server.BaseChangeCo
     }
   }
 
-  protected static boolean schemaHasField(Schema schema, String field) {
+  protected static boolean schemaContainsField(Schema schema, String field) {
     try {
       if (schema.getFields() != null && schema.getFields().get(field) != null) {
         return true;

--- a/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/BaseRecordConverter.java
+++ b/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/BaseRecordConverter.java
@@ -192,7 +192,11 @@ public abstract class BaseRecordConverter implements RecordConverter {
 
     ArrayList<String> keyFields = this.keyFields();
     if (keyFields.isEmpty()) {
+      if (value().has(clusteringField)) {
       return Clustering.newBuilder().setFields(List.of(clusteringField)).build();
+      } else {
+        return null;
+      }
     } else {
       // NOTE Limit clustering fields to 4. it's the limit of Bigquery
       List<String> clusteringFields = keyFields.stream().limit(3).collect(Collectors.toList());

--- a/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/BatchBigqueryChangeConsumer.java
+++ b/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/BatchBigqueryChangeConsumer.java
@@ -105,7 +105,7 @@ public class BatchBigqueryChangeConsumer<T> extends BaseChangeConsumer {
           .setMaxBadRecords(0)
           .setCreateSession(true);
 
-      if (schema.getFields() != null && schema.getFields().get(timePartitioning.getField()) != null) {
+      if (schemaHasField(schema, timePartitioning.getField())) {
         wCCBuilder.setTimePartitioning(timePartitioning);
       }
 

--- a/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/BatchBigqueryChangeConsumer.java
+++ b/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/BatchBigqueryChangeConsumer.java
@@ -105,10 +105,12 @@ public class BatchBigqueryChangeConsumer<T> extends BaseChangeConsumer {
           .setMaxBadRecords(0)
           .setCreateSession(true);
 
-      if (!config.common().nestedAsJson()) {
-        wCCBuilder
-            .setTimePartitioning(timePartitioning)
-            .setClustering(clustering);
+      if (schema.getFields() != null && schema.getFields().get(timePartitioning.getField()) != null) {
+        wCCBuilder.setTimePartitioning(timePartitioning);
+      }
+
+      if (clustering != null && clustering.getFields() != null && !clustering.getFields().isEmpty()) {
+        wCCBuilder.setClustering(clustering);
       }
 
       //WriteChannel implementation to stream data into a BigQuery table. 

--- a/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/BatchBigqueryChangeConsumer.java
+++ b/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/BatchBigqueryChangeConsumer.java
@@ -105,7 +105,7 @@ public class BatchBigqueryChangeConsumer<T> extends BaseChangeConsumer {
           .setMaxBadRecords(0)
           .setCreateSession(true);
 
-      if (schemaHasField(schema, timePartitioning.getField())) {
+      if (schemaContainsField(schema, timePartitioning.getField())) {
         wCCBuilder.setTimePartitioning(timePartitioning);
       }
 

--- a/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/StreamBigqueryChangeConsumer.java
+++ b/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/StreamBigqueryChangeConsumer.java
@@ -239,7 +239,7 @@ public class StreamBigqueryChangeConsumer extends BaseChangeConsumer {
     if (schemaContainsField(schema, timePartitioning.getField())) {
       tableDefBuilder.setTimePartitioning(timePartitioning);
     } else {
-      LOGGER.warn("PTT field `{}` not found in table schema, creating {} without PTT definition", timePartitioning.getField(), tableId);
+      LOGGER.warn("PTT field `{}` not found in table schema, creating {} without PTT definition", timePartitioning.getField(), tableId.getTable());
     }
 
     if (tableConstraints != null) {

--- a/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/StreamBigqueryChangeConsumer.java
+++ b/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/StreamBigqueryChangeConsumer.java
@@ -238,6 +238,8 @@ public class StreamBigqueryChangeConsumer extends BaseChangeConsumer {
 
     if (schemaContainsField(schema, timePartitioning.getField())) {
       tableDefBuilder.setTimePartitioning(timePartitioning);
+    } else {
+      LOGGER.warn("PTT field `{}` not found in table schema, creating {} without PTT definition", timePartitioning.getField(), tableId);
     }
 
     if (tableConstraints != null) {

--- a/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/StreamBigqueryChangeConsumer.java
+++ b/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/StreamBigqueryChangeConsumer.java
@@ -236,7 +236,7 @@ public class StreamBigqueryChangeConsumer extends BaseChangeConsumer {
 
     StandardTableDefinition.Builder tableDefBuilder = StandardTableDefinition.newBuilder().setSchema(schema);
 
-    if (schemaHasField(schema, timePartitioning.getField())) {
+    if (schemaContainsField(schema, timePartitioning.getField())) {
       tableDefBuilder.setTimePartitioning(timePartitioning);
     }
 

--- a/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/StreamBigqueryChangeConsumer.java
+++ b/debezium-server-bigquery-sinks/src/main/java/io/debezium/server/bigquery/StreamBigqueryChangeConsumer.java
@@ -236,7 +236,7 @@ public class StreamBigqueryChangeConsumer extends BaseChangeConsumer {
 
     StandardTableDefinition.Builder tableDefBuilder = StandardTableDefinition.newBuilder().setSchema(schema);
 
-    if (!config.common().nestedAsJson()) {
+    if (schemaHasField(schema, timePartitioning.getField())) {
       tableDefBuilder.setTimePartitioning(timePartitioning);
     }
 

